### ISSUE TITLE
harvester ID

### ISSUE THREAD

**Issue by philippeluickx** [apinf/api-harvestor#3]
We need to keep track of the APIs we harvest somehow. This is important for when we want to update APIs. You can not simply add an API on every run.
If you want to update an API on apinf.io, you need to be able to identify that API on both sides (harvester and apinf.io)

**Comment by kyyberi** [comment]
That's where we need to use catalog REST API to get list or search for an API before pushing or updating an API. Based on the results, our app desides is the current API new or is this just update and whether anything has changed.

**Comment by philippeluickx** [comment]
Right, but we still need to keep an ID somewhere that is unique that points back to the harvester and the source. Otherwise we won't be able to differentiate between an API manually added or the same API harvested.

That also brings a more philosophical issue: how do we deal with duplicate APIs? Anyone can add Google maps API. Should we group all APIs that point to the same API endpoint somehow?

**Comment by bajiat** [comment]
I'm worried about the philosophical issue mentioned by @philippeluickx If we are harvesting from several sources, each and everyone of them could have the same very popular API (like the Google maps you mentioned). Do we add it multiple times? Treat it as a duplicate?  (And how is it recognised as a duplicate?) Which is the primary source in that case?

**Comment by philippeluickx** [comment]
From a design perspective, all these duplicates are horrible. As long as there is no single, verified owner for an API, it is possible for any user to add more duplicates.
Not quite there yet, but I suggest the following at least:
- If there is an official owner of the API, we mark it somehow and we won't allow anyone else to add this API.
-- Any previously defined API (in case the owner shows up later) will not be accessible through the webservice, but we do forward the calls from old APIs to the owned one.
-- The owner will get the possibility to migrate existing data (curated content) to his/her owned API.
- If there is no official owner yet, anyone can add this same API again. From a design perspective, we do group all the API owners together and indicate this in the UI (multiple owners). Here the rating will become important, as the default shown should be the one with the highest rating then.
-- The content of each API might be different... so all of them should be accessible.

In any case, worries for later. Let's solve this once it comes up.

**Comment by kyyberi** [comment]
I understand philosophical concerns but atm we are collecting information. It is different process that decides what information is added to catalog. This issue will not stop the process now. If you find a solution to it now, good.

**Comment by kyyberi** [comment]
What we are discussing now is Single Source of Truth. That does not exist nowadays anywhere (easily) and aiming for it is the source of trouble https://rvsoapbox.blogspot.fi/2010/03/whats-wrong-with-single-version-of.html & https://www.linkedin.com/pulse/master-data-management-mdm-single-source-truth-jim-mckinney

SSoT has been one of the biggest issues in industry for years. You can achieve it, but what is the cost and does it have the desired effect? SSoT is needed when we have critical data. This is not critical data IMHO.

**Comment by phanimahesh** [comment]
I'm jumping on the wagon midway, may have gotten the problem statement/context not quite right, but...

I will have to disagree with the comment on single source of truth. Aiming for single source of truth can become limiting and troublesome in certain cases, while being incredibly useful or even essential in other cases. I am not convinced that the principle itself is at fault. Imagine if bank account balances did not have a single source of truth/single version of truth. There can be internal inconsistencies (happens in real life) so long as there is a well defined protocol to resolve them. In the case of bank records the protocol can be balance recomputations from account ledgers, and verification of ledger entries.

It is difficult to come up with a source when data is not in a good, universally agreed, structured format. One possible solution is to allow duplicates (potentially allowing inconsistent/different data representing the same entity), but to define a resolution protocol.

Here is a protocol proposal:

 - For each new api, if entries exist with the same base url, and if a verified entry exists, it will be considered canonical. We can optionally allow the new entry to be made, while adding internal metadata that this refers to that canonical api.
 - Whenever an api is verified, a search is performed to make sure that all existing duplicates have this metadata applied to them.
 - If no verified api exists matching the base url, a new entry is allowed. If the data entered matches any existing one, either exactly or approximately, a confirmation may be requested for manually made entries.

The process of verification is unspecified as of now, and is open to discussion. Thoughts?

**Comment by kyyberi** [comment]
Sounds valid to me. Thanks for jumping in!

**Comment by philippeluickx** [comment]
Valid points.
For me the main concern is "ownership". How do we keep the owners of a (verified) API happy? Plus, typically the owner of information has always more authority.
There is a discussion going on about what content should be "owned" by an API owner and which content can be community-curated. It's a bit like Github meets Stackoverflow.
With Github, you typically go to the main branch and most of the times ignore the forks. Main branch is the owner of the repo and most often you respect that and try to centralize all efforts there.
With Stackoverflow though, the whole drive comes from the community and if the author of the code cares a bit, the author can jump in (but it's not required).

We'll have to keep an open mind when moving forward, make some choices and gather feedback. Definitely more opinions required on this discussion!
